internal/p2p/peermanager.go

```diff
@@ -1287,15 +1289,22 @@ func (p *peerInfo) Score() PeerScore {
 		return PeerScorePersistent
 	}
 
-	if p.MutableScore <= 0 {
+	score := p.MutableScore
+	for _, addr := range p.AddressInfo {
+		// DialFailures is reset when dials succeed, so this
+		// is either the number of dial failures or 0.
+		score -= int64(addr.DialFailures)
+	}
```

Comment:
This is more of a nit as there is nothing wrong with the logic per say, I just don't think this logic should be added in the `Score()` method but rather should alter `p.MutableScore` when `DialFailed` is called. I think it would make it easier for readers of the code to understand what is happening that way.

Reply:
The reason to have it here is that `DialFailures` is reset when dialing succeeds. If we do it the other way (which I tried initially) we end up penalizing a peer for having connectivity problems for a long time after its recovered (if it's recovered,) which might lead to us evicting a peer that's in a good state.

Reply:
It is possible to counter this by adding back `addr.DialFailures` to the score before resetting it when dialing succeeds.

Reply:
Huh? but adding back `addr.DialFailures` to the score will offset the previous penalties for not being able to connect.

Reply:
I pushed a change that I think resolves this.

Reply:
I had something slightly different in mind but this is good enough